example-client/README.md

```diff
@@ -3,11 +3,219 @@ Example CoAP client
+
+```
+openssl req -key my_ca_private_key.pem -new -sha256 -x509 -days 12775 -out my_ca_public_cert.pem -subj /CN=CA -config <(echo '[req]'; echo 'distinguished_name=dn'; echo '[dn]'; echo '[ext]'; echo 'basicConstraints=CA:TRUE') -extensions ext
+```
+## LWM2M Device Certificates
+
+### Create device private key
+
+```
+openssl ecparam -out lwm2m_device_private_key.pem -name prime256v1 -genkey
+```
+
+### Convert private key to DER format
+
+```
+openssl ec -in lwm2m_device_private_key.pem -out lwm2m_device_private_key.der -outform der
+```
+
+### Create a certificate signing request (CSR) for the private key
+
+```
+openssl req -key lwm2m_device_private_key.pem -new -sha256 -out lwm2m_device_private_key_csr.pem -subj /CN=${ENDPOINT_NAME}
+```
+
+### Sign the certificate signing request (CSR) with the CA private key and certificate -
+
+```
+openssl x509 -req -in lwm2m_device_private_key_csr.pem -sha256 -out lwm2m_device_cert.der -outform der -out lwm2m_device_cert.pem -outform pem -CA my_ca_public_cert.pem -CAkey my_ca_private_key.pem -CAcreateserial -days 3650
+```
+
+## Java Keystore
+
+#### Create a PKCS #12 archive to store cryptography objects so we can create a Java keystore from it
+
+```
+openssl pkcs12 -export -in lwm2m_device_cert.pem -inkey lwm2m_device_private_key.pem -out my_lwm2m_device_keystore_certs.p12 -name mbed-cloud-${ENDPOINT_NAME} -password pass:secret
+```
+
+#### Create Java keystore importing the PKCS #12 archive
+
+```
+keytool -keystore example-client-device-keystore.jks -storepass secret -importkeystore -srcstorepass secret -srckeystore my_lwm2m_device_keystore_certs.p12 -srcstoretype PKCS12 -noprompt
+```
+
+#### Import our CA cert we created above to the Java keystore
+
+```
+keytool -keystore example-client-device-keystore.jks -storepass secret -import -alias lwm2m_server_ca_certificate -file lwm2m_server_ca_certificate.pem -noprompt
+```
+
+## Convience Commands Viewing/Verifying Keys, Certs, Keystore
+
+### Dump the contents of the certs we created for convience
+
+```
+openssl pkcs12 -in my_lwm2m_device_keystore_certs.p12 -nodes -passin pass:secret | openssl x509 -noout -subject
+openssl x509 -in my_ca_public_cert.pem -noout -text | grep Subject:
+openssl x509 -in lwm2m_server_ca_certificate.pem -noout -text | grep Subject:
+openssl x509 -in lwm2m_device_cert.der -inform der -text -noout
+```
+
+### Examine the keystore
+
+```
+keytool -list -v -storepass secret -keystore example-client-device-keystore.jks
+```
+
+## Mbed Cloud Certificate Uploading
+
+### Output a hash of our private key plus account nexus in base64 for mbedcloud verification
+
+Note: You only need this if you are uploading the cert with using the Portal UI
+
+```
+printf "${ACCOUNT_ID}" > account_id && openssl dgst -sha256 -sign my_ca_private_key.pem -out sign.sha256 account_id && base64 sign.sha256
+```
+
+### Get a json payload file to upload the results with cURL if you didn't do this via Portal
+
```

Comment:
- Again, a very long heading. Avoid these in general. Keep them short and explain the beef in the text under the heading.
- Portal -> Mbed Cloud portal